enkit/machinecert/install.go

```diff
@@ -0,0 +1,156 @@
+package machinecert
+
+import (
+	"context"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/enfabrica/enkit/astore/rpc/auth"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"golang.org/x/crypto/ssh"
+)
+
+func (i *Install) Run(cmd *cobra.Command, args []string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// Get connection to auth server
+	conn, err := i.root.Connect()
+	if err != nil {
+		return fmt.Errorf("can't connect to auth server: %w", err)
+	}
+	authClient := auth.NewAuthClient(conn)
+
+	// Check file overwriting
+	_, overwritten, err := i.getAffectedFiles()
+	if err != nil {
+		return fmt.Errorf("while determining changed files: %w", err)
+	}
+	if len(overwritten) > 0 && !i.Overwrite {
+		return fmt.Errorf("--overwrite not specified, but command would change files: %v", overwritten)
+	}
+
+	privateKey, err := readFileOrEmpty(i.ExistingPrivateKeyPath)
+	if err != nil {
+		return fmt.Errorf("while loading private key: %w", err)
+	}
+	publicKey, err := readFileOrEmpty(i.ExistingPublicKeyPath)
+	if err != nil {
+		return fmt.Errorf("while loading public key: %w", err)
+	}
+
+	if (publicKey != nil && privateKey == nil) || (publicKey == nil && privateKey != nil) {
+		return fmt.Errorf("If one of [--existing-private-key, --existing-public-key] is supplied, both must be supplied")
+	}
+	if privateKey == nil || publicKey == nil {
+		// Create new keypair
+		pubKey, privKey, err := kcerts.GenerateED25519()
+		if err != nil {
+			return fmt.Errorf("failed to generate new keypair: %w", err)
+		}
+		publicKey = pubKey.Marshal()
+		publicKeyAuthLine := ssh.MarshalAuthorizedKey(pubKey)
+
+		// Write private key to disk
+		privateKey, err = privKey.SSHPemEncode()
+		if err != nil {
+			return fmt.Errorf("while encoding private key: %w", err)
+		}
+		if err := ioutil.WriteFile(i.root.PrivateKeyPath, privateKey, 0400); err != nil {
```

Comment:
All perms are `0400` which isn't super convenient; need to double-check what they should be